components/ca-form.js

```diff
@@ -0,0 +1,1521 @@
+                    itemIndex++;
+
+                    // keep a count of the number of items in this column
+                    itemsPerCol++;
+
+                    // add input element to the elements collection so we can keep a track of all editable children
+                    this.elements.push(inputEl);
+                }
+
+                // disable all child controls if the form is readonly
+                if (this.readonly) {
+
+                    const inputs = Array.prototype.slice.call(this.form.querySelectorAll('input,select,textarea'));
+
+                    inputs.forEach(item => {
+                        item.setAttribute('disabled', 'true');
+                    });
+                }
+
+                // if the user can create, render a save button
+                const createLink = schema.links && schema.links.find(link => link.rel === 'create');
+                let resetButton = null;
+                let buttonContainer = null;
+
+                if (createLink) {
+
+                    // set create/update urls
+                    this.createUrl = createLink.href;
+                    this.createMethod = createLink.method || 'POST';
+
+                    const createSchema = createLink.properties;
+
+                    // if we have a schema of items, add them to the url
+                    // TODO: rewrite this to better handle schemas (hacked in for the demo)
+                    // We can now handle parameterized schemas on ca-list for instances, so make use of that.
+                    if (createSchema) {
+                        // eslint-disable-next-line no-restricted-syntax
+                        for (const key in createSchema) {
+
+                            if (Object.prototype.hasOwnProperty.call(createSchema, key) && this.createUrl.indexOf(`{${key}}`)) {
+
+                                const schemaItem = createSchema[key];
+
+                                // if we have a required guid, add one
+                                if (schemaItem.required && schemaItem.format === 'uuid') {
+                                    this.createUrl = this.createUrl.replace(`{${key}}`, newGuid());
+                                }
+                            }
+                        }
+                    }
+
+                    // insert button container
+                    buttonContainer = createElementLegacy(this.form, 'div', { class: 'buttons' });
+
+                    // add submit button
+                    createElementLegacy(buttonContainer, 'input', { type: 'submit', value: createLink.title || 'Save' });
+
+                    // add reset button if we have it.
+                    if (this.reset) {
+                        resetButton = createElementLegacy(buttonContainer, 'input', { name: 'reset', type: 'button', value: 'Reset' });
+                        resetButton.onclick = () => {
+                            this.src = this.src;
```

Comment:
yes it does look odd. Perhaps if the object that gets passed as a parameter was assigned to a variable it would be more readable...